File: cpp/zero-array-transformation-ii.cpp
class Solution {
private:
    // function to check if the queries till mid value will be able to make the nums vector zero
    bool isRight(vector<int>& nums,vector<int> &prefix) {
        for (int i=0;i<nums.size();i++) {
            int val = nums[i]-prefix[i];
            if (val>0) {
                return false;
            }
        }
        return true;
    }
public:
    int minZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int l = 0, r = queries.size();
        int ans = -1;
        // binary search to find the mid value and then check whether the queries till there will be able to get a zero-array
        while (l<=r) {
            int mid = l + (r-l)/2;
            // prefix sum to efficiently calculate the changes till ith query
            // the first is incremented by the val and the last index + 1 is decremented by the val so that while calculating prefix sum the range is incremented by k and no other value is affected since the -val at last index + 1 will make for it
            vector<int> prefix(nums.size()+1,0);
            for (int i=0;i<mid;i++) {
                prefix[queries[i][0]]+=queries[i][2];
                prefix[queries[i][1] + 1]-=queries[i][2];
            }
            for (int i=1;i<prefix.size();i++) {
                prefix[i]+=prefix[i-1];
            }
            if (isRight(nums,prefix)) {
                ans = mid;
                r = mid-1;
            }
            else {
                l = mid+1;
            }
        }
        return ans;
    }
};

File: java/zero-array-transformation-ii.java
class Solution {
    public int minZeroArray(int[] nums, int[][] queries) {
        int sum = 0, k = 0;
        int n = nums.length;
        // prefix sum to calculate the changes using queries efficiently by increasing it with val at max(l,i) and decrementing at (r+1) so that the prefix is balanced after r and changes are made only in the range l to r 
        int prefix[] = new int[n+1];
        // for every index checking whether the prefix sum that is sum at the current index is sufficient or not
        for (int i=0;i<n;i++) {
            while ((sum+prefix[i])<nums[i]) {
                // queries weren't sufficient so return -1
                if (k==queries.length) return -1;
                int l = queries[k][0],r = queries[k][1], val = queries[k][2];
                k++;
                if (r<i) continue;
                // changing the prefix sum as stated above
                prefix[Math.max(l,i)]+=val;
                prefix[r+1]-=val;
            }
            // making changes to sum to check for every index is sufficient or not
            sum+=prefix[i];
        }
        return k;
    }
}
